### PR TITLE
feat: add master config file

### LINKED ISSUE #485: Config file for nfd-master
**What would you like to be added**:

Now that nfd-worker has quite extensive support for dynamic run-time configurability through its config file, I started to think would it be good to have something similar for nfd-master as well 🤔  A similar json/yaml based configuration file with fsnotify watching for changes. Could allow dynamic control of e.g. logging parameters, allowed namespaces, extended resources and label whitelisting.

**Why is this needed**:

Would make configuration of nfd-master easier, enabling dynamic (run-time) configurability.

### PR DESCRIPTION
**Pull request by AhmedGrati**
Resolves #485.
Tasks done:
- [x] Support config file for the master, with `fsnotify`, `config parsing`, and `overrides from command line flags`
- [x] Unit Tests
- [x] Deployment (Kustomize, Helm, Update `make templates`) with the default configuration
- [x] Documentation (command line reference, config reference, nfd-master usage page)
- [x] E2E Tests

### CODE REVIEW

**Comment by marquiz** [comment]
Thanks @AhmedGrati for the enhancement! I will look into this shortly
/ok-to-test
/assign @fmuyassarov  @ArangoGutierrez  @marquiz

File: docs/usage/nfd-master.md
Comment (AhmedGrati) [reply docs/usage/nfd-master.md:77]: I'll take care of it in a separate PR.

File: pkg/nfd-master/nfd-master.go
Comment (AhmedGrati) [reply pkg/nfd-master/nfd-master.go:1016]: Yes that's gonna be helpful! I'll do it.

File: deployment/components/master-config/nfd-master.conf.example
Comment (AhmedGrati) [reply deployment/components/master-config/nfd-master.conf.example:2]: @marquiz the same would be applied to denyLabelNs doesn't it?

File: pkg/nfd-master/nfd-master.go
Comment (marquiz) [reply pkg/nfd-master/nfd-master.go:1016]: 👍 This is something we might want to add to nfd-worker too (in a separate PR), if we don't have it

File: deployment/components/master-config/nfd-master.conf.example
Comment (marquiz) [reply deployment/components/master-config/nfd-master.conf.example:2]: Yeah, that applies to all of the `StringSetVal` options i.e. those two plus `ResourceLabels`

Comment (AhmedGrati) [reply deployment/components/master-config/nfd-master.conf.example:2]: I'll wait until your PR gets merged.

File: deployment/base/master/master-deployment.yaml
Comment (marquiz) [inline deployment/base/master/master-deployment.yaml:37]: nit: stale empty line added, no need to touch this file

File: docs/reference/master-configuration-reference.md
Comment (marquiz) [inline docs/reference/master-configuration-reference.md:55]: Just an array, or list. In yaml they are called array but would list be more descriptive for the user. WDYT?

```suggestion
`denyLabelNs` specifies a list of excluded
```

File: docs/reference/master-configuration-reference.md
Comment (marquiz) [inline docs/reference/master-configuration-reference.md:70]: just a note that we need to fix all of these when #1087 gets merged and we rebase this
```suggestion
denyLabelNs: ["denied.ns.io","denied.kubernetes.io"]
```

File: docs/usage/nfd-master.md
Comment (marquiz) [inline docs/usage/nfd-master.md:77]: Did I earlier suggest a change that broke this(?) 🙈

```suggestion
[Master pod parameter](../deployment/helm.md#master-pod-parameters)
```

**Comment by fmuyassarov** [comment]
Thanks for working on this @AhmedGrati , I will give some time for the review by the end of the day.

File: test/e2e/node_feature_discovery_test.go
Comment (marquiz) [inline test/e2e/node_feature_discovery_test.go:814]: I think we'd like to verify that the config file was actually taken into effect by nfd-master. We don't necessarily need to verify the annotations here as it's done already in other tests (just to speed up things a bit).

What we could do, for example
1. deploy a NodeFeature object to create some features on one worker node
2. deploy NodeFeatureRule to taint the node
3. Verify that the expected taint was created
4. Modify the configmap to `enableTaints: false``
5. Verify that the taint was removed from the node

For the NodeFeature object you could probably just re-use `test/e2e/data/nodefeature-1.yaml`. For the NodeFeatureRule you could create something new, simple like
```yaml
spec:
  rules:
    - name: "e2e-tainting-test"
      taints:
        - effect: PreferNoSchedule
          key: "nfd.node.kubernetes.io/e2e-test"
      matchFeatures:
        - feature: "e2e.flags"
          matchExpressions:
            "flag_1": {op: Exists}
```

```suggestion
```

Comment (marquiz) [reply test/e2e/node_feature_discovery_test.go:814]: You can take a look at other tests, e.g. those for taints for some inspiration how the taints are created and verified

File: test/e2e/node_feature_discovery_test.go
Comment (marquiz) [inline test/e2e/node_feature_discovery_test.go:851]: We shouldn't repeat all of this setup teardown stuff. What you could do is e.g. something similar to what topologyupdater tests (`test/e2e/topology_updater_test.go`) do, combining `JustBeforeEach` and `BeforeEach`. Basically use `BeforeEach` in each test to set the exact parameters of nfd-master.

**Review by marquiz — COMMENTED** [review]
Thanks for the update @AhmedGrati, starts to look really good. Re-labeling on updates (when NodeFeature API is enabled) needs to be added. Also, a bit sanitizing of the e2e part.

One additional ask: it would be really cool to get the klog parameters to the config file as well, similar to nfd-worker. That can be done in a separate PR, though, I think.

File: test/e2e/node_feature_discovery_test.go
Comment (marquiz) [inline test/e2e/node_feature_discovery_test.go:786]: What I tried to express is that we don't need to repeat all the set-up tear down stuff that's already present in the test suite above. We just need to change [this](https://github.com/kubernetes-sigs/node-feature-discovery/blob/13f92faa771d58c15b40e92543cdef8bd711021e/test/e2e/node_feature_discovery_test.go#L206) to `JustBeforeEach` and adjust the tests with a `BeforeEach` to create the nfd-master pod with correct options.

Move this test to the suite above and we can drop basically all the BeforeAll, AfterAll, JustBeforeEach etc.

File: docs/reference/master-configuration-reference.md
Comment (fmuyassarov) [inline docs/reference/master-configuration-reference.md:26]: can we add here taints too?

**Review by fmuyassarov — COMMENTED** [review]
Looks good to me. Thanks @marquiz for helping reviewing and @AhmedGrati for the all work here.

File: docs/reference/master-configuration-reference.md
Comment (AhmedGrati) [reply docs/reference/master-configuration-reference.md:26]: @fmuyassarov do you mean like that?
```suggestion
`noPublish` option disables updates to the Node objects in the Kubernetes
API server, making a "dry-run" flag for nfd-master. No Labels, Annotations, Taints
or ExtendedResources of nodes are updated.
```

File: test/e2e/node_feature_discovery_test.go
Comment (marquiz) [inline test/e2e/node_feature_discovery_test.go:177]: I'd suggest to make this more generic e.g. with smth like

```suggestion
				customMasterPodSpecOpts              =  *[]testpod.SpecOption
```

With this the test can customize the nfd-master pod as they like. Use the default opts if `customMasterPodSpecOpts == nil`, otherwise use the spec opts provided by it. WDYT?

Comment (AhmedGrati) [reply test/e2e/node_feature_discovery_test.go:177]: @marquiz the configmap key and name are needed to create a new configmap, which is unrelated to the master deployment, so we can't put them there. wdyt?

Comment (marquiz) [reply test/e2e/node_feature_discovery_test.go:177]: > the configmap key and name are needed to create a new configmap, which is unrelated to the master deployment, so we can't put them there. wdyt?

Yup, but you should be able to create the configmap in the BeforeEach() function of the specific test that needs it, and set the PodSpecOpts accordingly, right?

Comment (AhmedGrati) [reply test/e2e/node_feature_discovery_test.go:177]: Yes, we are already doing that in the `JustBeforeEach`, but I think we still need the configMapName and configMapKey even outside the `JustBeforeEach` clause, that's why we need it in the `var` section.

Comment (marquiz) [reply test/e2e/node_feature_discovery_test.go:177]: You shouldn't need those. You should be able to configure that with the `SpecWithConfigMap()` podSpecOpt(?)

Comment (marquiz) [reply test/e2e/node_feature_discovery_test.go:177]: You wouldn't need to have those as "top level" variables. Just hard code them (as you know the names in the test) or specify variables on the inner Context-.

File: test/e2e/node_feature_discovery_test.go
Comment (marquiz) [inline test/e2e/node_feature_discovery_test.go:677]: This should be in the `BeforeEach()` of the new test instead

File: test/e2e/node_feature_discovery_test.go
Comment (marquiz) [inline test/e2e/node_feature_discovery_test.go:797]: This should be added into the "main" AfterEach() method instead. I.e. the one that where we cleanup the labels and CRs.

File: test/e2e/node_feature_discovery_test.go
Comment (marquiz) [inline test/e2e/node_feature_discovery_test.go:241]: These can and should be done in the BeforeEach() method of the test that needs the configmap

**Review by marquiz — COMMENTED** [review]
Thanks @AhmedGrati for the update! I think we're almost there, just a few more suggestions

**Review by marquiz — COMMENTED** [review]
Nice work @AhmedGrati, for me the code looks good 👍

However, I have one more wish 😇 I only now looked at the commit message which is very brief (there is no message "body" whatsoever). Could you add a bit more flesh to it, explaining what is being added, see [k8s commit msg guidelines](https://www.kubernetes.dev/docs/guide/pull-requests/#use-the-commit-message-body-to-explain-the-_what_-and-_why_-of-the-commit)

**Comment by AhmedGrati** [comment]
@marquiz done.

**Review by marquiz — APPROVED** [review]
Thanks @AhmedGrati! To me this now looks good to go

**Comment by marquiz** [comment]
@ArangoGutierrez @zvonkok wanna check this one?

**Review by ArangoGutierrez — APPROVED** [review]
/lgtm

**Pull request merged**
